src/main/java/com/getindata/connectors/http/internal/auth/OidcAccessTokenManager.java

```diff
@@ -0,0 +1,148 @@
+        this.httpClient = httpClient;
+        this.url = url;
+        if (null == tokenExpirationReduction) {
+            this.tokenExpirationReduction = DEFAULT_TOKEN_EXPIRATION_REDUCTION;
+        } else {
+            this.tokenExpirationReduction = tokenExpirationReduction;
+        }
+    }
+
+    /**
+     * Request an access token from the token endpoint
+     */
+    private void requestAccessToken() {
+        try {
+            HttpRequest httpRequest =
+                    HttpRequest.newBuilder()
+                    .uri(URI.create(url))
+                    .header("Content-Type", "application/x-www-form-urlencoded")
+                    .method("POST", HttpRequest.BodyPublishers.ofString(tokenRequest))
+                    .build();
+
+            HttpResponse<byte[]> response = httpClient.send(httpRequest,
+                    HttpResponse.BodyHandlers.ofByteArray());
+            //create ObjectMapper instance
+            ObjectMapper objectMapper = new ObjectMapper();
+            if (200 == response.statusCode()) {
+                byte[] bytes = response.body();
+                JsonNode rootNode = objectMapper.readTree(bytes);
+                JsonNode tokenNode = rootNode.path("access_token");
+                JsonNode expiresInNode = rootNode.path("expires_in");
+                this.cachedAccessToken = tokenNode.textValue();
+                /*
+                 expiresIn is in seconds
+                */
+                Duration expiresIn = Duration.ofSeconds(expiresInNode.asInt());
+                if (expiresIn.compareTo(this.tokenExpirationReduction) > 0) {
+                    //expiresIn is greater than tokenExpirationReduction
+                    expiresIn = expiresIn.minus(this.tokenExpirationReduction);
+                }
+                this.cachedAccessTokenExp = Instant.now().plus(expiresIn);
+            } else {
+                throw new IllegalStateException("Attempted to get an access token but got http" +
+                        " status code " + response.statusCode());
+            }
+        } catch (JsonProcessingException e) {
+            throw new IllegalStateException("Error found while trying to request a new token");
+        } catch (IOException e) {
+            throw new IllegalStateException("IO Exception occurred",  e);
+        } catch (InterruptedException e) {
+            throw new IllegalStateException("Interrupted Exception occurred", e);
+        }
+    }
+
+    /**
+     * Get a valid unexpired access token.
+     * @return access token.
+     */
+    public String authenticate() {
+        if (isAccessTokenRequired()) {
+            requestAccessToken();
+        }
+        return cachedAccessToken;
+    }
+
+    private boolean isAccessTokenRequired() {
+        return null == cachedAccessToken || isTokenExpired();
+    }
+
+    private boolean isTokenExpired() {
+        return Instant.now().isAfter(this.cachedAccessTokenExp);
+    }
+}
```

Comment:
There is a lot of new base code, and at the same time I dont see any tests that will use this code. I think test coverage will fail on this.

Regardless test coverage, can you add tests that will use/verify this code?
Can something be mocked using WireMoc?

Reply:
@kristoffSC good idea - I will have a look.

Reply:
@kristoffSC is this the sort of thing you meant?